### PR TITLE
Migrate field ref-based column keys in `column_settings`

### PR DESCRIPTION
**Pull request by ranquild**
Related to https://github.com/metabase/metabase/pull/46383

We had a lot of migrations to maintain field refs in `column_settings` in the past, and I hope this is the last migration.

Currently, we have 2 column keys used in `column_settings` - field ref-based (`["ref", ["field", 1, null]]`) and name-based (`["name", "ID"]`). We use(d) field-ref keys for simple cases where a query is based on a raw table, and name-based in more complex cases (this is simplified). This was error prone because of the following reasons:
- Makes it very hard to do any MBQL changes/fixes because every change would require a custom migration
- Field ref-based keys omit `:binning` and `temporal-unit` options, making it impossible to have columns that differ only by this options
- Some changes to a query (new stage etc) would trigger field ref changes and loss of all settings
- Any bugs with field refs (self joins etc) automatically result in bugs in viz settings

This PR migrates all `column_settings` keys to name-based keys. The FE in the referenced PR supports both field ref- and name-based keys for the keys that would not be migrated by this migration (no `result_metadata`).

How to verify:
- Switch to the base branch (master)
- Create questions with `column_settings`. These are formatting settings for columns. E.g. New -> Orders -> Visualize, click on column headers -> Gear icon -> Change column title -> Save question
- Checkout this branch and run
- All your viz settings should be migrated. There should be no visual changes. Export should continue working and it should use the updated settings.
- You can check that the settings are migrated by inspecting Network and making sure that `["name", ...]` keys are used in `column_settings` in `visualization_settings`.

### CODE REVIEW

File: frontend/src/metabase-types/api/card.ts [hunk 1]
```diff
@@ -126,13 +126,21 @@ export type XAxisScale = "ordinal" | "histogram" | "timeseries" | NumericScale;
 
 export type YAxisScale = NumericScale;
 
+export interface ColumnSetting {
+  column_title?: string;
+
+  // some options are untyped
+  [key: string]: any;
```
Comment (kamilmielnik) [inline frontend/src/metabase-types/api/card.ts:133]:
```suggestion
  [key: string]: unknown;
```

File: e2e/test/scenarios/dashboard-filters/temporal-unit-parameters.cy.spec.js [hunk 2]
```diff
@@ -387,9 +387,9 @@ describe("scenarios > dashboard > temporal unit parameters", () => {
         .should("contain.text", "Started from")
         .should("contain.text", multiBreakoutQuestionDetails.name);
       queryBuilderMain().within(() => {
-        cy.findByText("Product → Created At: Week").should("be.visible");
-        cy.findByText("2022").should("be.visible");
-        cy.findByText("2023").should("be.visible");
+        cy.findByText("Created At: Year").should("be.visible");
```
Comment (kamilmielnik) [inline e2e/test/scenarios/dashboard-filters/temporal-unit-parameters.cy.spec.js:390]: Before this PR: default pivot column is "Created At": [image](https://github.com/user-attachments/assets/025ff332-a337-46bf-b563-0bc78b16b846)

After this PR: default pivot column is "Product -> Created At": [image](https://github.com/user-attachments/assets/4c27cc81-961c-46fd-89b6-9bea062cd514)

Arguably it was correct/better before so I'd say this is a regression.

**Review by kamilmielnik — COMMENTED** [review]
Looks good, works well :+1: 
I only reviewed the frontend changes though.

File: frontend/src/metabase-lib/v1/queries/utils/get-column-key.ts [hunk 3]
```diff
@@ -1,44 +1,10 @@
...
-  const isLegacyRef =
-    (isFieldReference(fieldRef) && hasStringFieldName(fieldRef)) ||
-    isAggregationReference(fieldRef);
-
-  return JSON.stringify(
-    isLegacyRef ? ["name", column.name] : ["ref", fieldRef],
-  );
-};
+export function getColumnNameFromKey(key: string): string | undefined {
+  const [type, name] = JSON.parse(key);
```
[the remark this answers is not in the document]
Comment (ranquild) [reply frontend/src/metabase-lib/v1/queries/utils/get-column-key.ts:8]: The BE guarantees that these keys can be parsed. If you try to save non-json the BE would throw

File: e2e/test/scenarios/dashboard-filters/temporal-unit-parameters.cy.spec.js (hunk 2 above)
Comment (ranquild) [reply e2e/test/scenarios/dashboard-filters/temporal-unit-parameters.cy.spec.js:390]: Nope this is the expected behavior, and there was a bug ticket to fix it. But I didn't realize that it was an issue with viz settings

**Review by calherries — COMMENTED** [review]
Wow, nice Clojure! What happens when a revision is restored for a Card or DashboardCard that has legacy column keys? Will that case be handled on read or something?

**Comment by ranquild** [comment]
@calherries that's correct. Both the FE and BE support legacy column keys and there are many existing tests for this behavior.

**Review by deniskaber — APPROVED** [review]

**Pull request merged**